cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java

```diff
@@ -162,30 +163,36 @@ public void removeTags(NamespacedEntityId namespacedEntityId, String... tags) th
   }
 
   @Override
-  public Set<MetadataSearchResultRecord> search(String namespaceId, String searchQuery,
-                                                Set<MetadataSearchTargetType> types,
-                                                SortInfo sortInfo) throws Exception {
-    return filterAuthorizedSearchResult(metadataStore.search(namespaceId, searchQuery, types, sortInfo));
+  public MetadataSearchResponse search(String namespaceId, String searchQuery,
+                                       Set<MetadataSearchTargetType> types,
+                                       SortInfo sortInfo, int offset, int limit,
+                                       int numCursors, String cursor) throws Exception {
+    return filterAuthorizedSearchResult(
+      metadataStore.search(namespaceId, searchQuery, types, sortInfo, offset, limit, numCursors, cursor)
+    );
   }
 
   /**
    * Filter a list of {@link MetadataSearchResultRecord} that ensures the logged-in user has a privilege on
    *
-   * @param results the {@link Set<MetadataSearchResultRecord>} to filter with
-   * @return filtered list of {@link MetadataSearchResultRecord}
+   * @param results the {@link MetadataSearchResponse} to filter
+   * @return filtered {@link MetadataSearchResponse}
    */
-  private Set<MetadataSearchResultRecord> filterAuthorizedSearchResult(Set<MetadataSearchResultRecord> results)
+  private MetadataSearchResponse filterAuthorizedSearchResult(MetadataSearchResponse results)
```

Comment:
good point. However, passing this to the store will not fix this problem, right? Since the Dataset itself does sorting + pagination when sort field and sort order are provided.

This might need more thought - should we handle it separately?